src/goto-programs/string_refine_preprocess.cpp

```diff
@@ -0,0 +1,1343 @@
+  const source_locationt &location,
+  const std::string &signature)
+{
+  assert(refined_string_typet::is_java_string_pointer_type(
+           function_type.return_type()));
+  dereference_exprt deref(lhs, lhs.type().subtype());
+  typet object_type=ns.follow(deref.type());
+  exprt object_size=size_of_expr(object_type, ns);
+  typet length_type, data_type;
+  get_data_and_length_type_of_string(deref, data_type, length_type);
+
+  std::string fnamel=id2string(function_name) + "_length";
+  std::string fnamed=id2string(function_name) + "_data";
+  declare_function(fnamel, length_type);
+  declare_function(fnamed, data_type);
+  function_application_exprt rhs_length;
+  function_application_exprt rhs_data;
+  rhs_length.type()=length_type;
+  rhs_data.type()=data_type.subtype();
+  rhs_length.function()=symbol_exprt(fnamel);
+  rhs_data.function()=symbol_exprt(fnamed);
+
+  exprt::operandst processed_arguments=process_arguments(
+    goto_program, i_it, arguments, location, signature);
+  rhs_length.arguments()=processed_arguments;
+  rhs_data.arguments()=processed_arguments;
+
+  symbol_exprt tmp_length=new_tmp_symbol("tmp_length", length_type);
+  symbol_exprt tmp_array=new_tmp_symbol("tmp_array", data_type.subtype());
+  member_exprt lhs_length(deref, "length", length_type);
+  member_exprt lhs_data(deref, "data", tmp_array.type());
+
+  // lhs=malloc(String *)
+
+  if(object_size.is_nil())
+    debug() << "string_refine_preprocesst::make_string_assign "
+            << "got nil object_size" << eom;
+
+  side_effect_exprt malloc_expr(ID_malloc);
+  malloc_expr.copy_to_operands(object_size);
+  malloc_expr.type()=pointer_typet(object_type);
+  malloc_expr.add_source_location()=location;
+
+  std::list<code_assignt> assigns;
+  assigns.emplace_back(lhs, malloc_expr);
+  assigns.emplace_back(tmp_length, rhs_length);
+  assigns.emplace_back(lhs_length, tmp_length);
+  assigns.emplace_back(tmp_array, rhs_data);
+  assigns.emplace_back(lhs_data, address_of_exprt(tmp_array));
+  insert_assignments(goto_program, i_it, assigns);
+}
+
+/*******************************************************************\
+
+Function: string_refine_preprocesst::make_assign
+
+  Inputs: a goto_program, a position in this program, an expression lhs,
+          a function type, a function name, a vector of arguments, a location
+          and a signature
+
+ Purpose: assign the result of the function application to lhs,
+          in case the function type is string, it does a special assignment
+          using `make_string_assign`
+
+\*******************************************************************/
+
+void string_refine_preprocesst::make_assign(
+  goto_programt &goto_program,
+  goto_programt::targett &i_it,
+  const exprt &lhs,
+  const code_typet &function_type,
+  const irep_idt &function_name,
+  const exprt::operandst &arg,
+  const source_locationt &loc,
+  const std::string &sig)
+{
+  debug() << "make assign " << function_name << eom;
```

Comment:
Just checking: There isn't a whole lot of debug output here - is that one intentionally left in?